### PR TITLE
Generalise the Console module

### PR DESCRIPTION
**Pull request by diml**
This is in preparation for #3068. This PR generalises the `Console` module. The console is now seen as an abstract system that can report messages and a status the user.

The module now allows to set the actual backend, compose backends, etc... #3068 will simply provide a new backend for the console.

### CODE REVIEW

File: src/stdune/console.ml [hunk 1]
```diff
@@ -1,78 +1,130 @@
...
-  let all =
-    [ ("progress", Progress)
-    ; ("verbose", Verbose)
-    ; ("short", Short)
-    ; ("quiet", Quiet)
-    ]
-end
+module Backend = struct
+  module type S = sig
+    val print : string -> unit
```
Comment (rgrinberg) [inline src/stdune/console.ml]: Not related to this PR, but why do we need this function? It seems like `Pp` should be encouraged here instead.

Comment (diml) [reply src/stdune/console.ml]: Yh, that's true actually. Let me have a look at this.

Comment (diml) [reply src/stdune/console.ml]: Using `Pp.t` here is indeed better and allows to improve the code and formatting in other places, such as in `Report_error`, but it's a bigger change so I'll submit a separate PR.

Comment (diml) [reply src/stdune/console.ml]: Done in #3242

**Pull request merged**